### DISCUSSION TITLE
Is it possible to mint NFT using eopsin.

### DISCUSSION THREAD

**Discussion by reitcircles** [OpShin/opshin#50] [Q&A] [answered]
We have the following requirement. For our NFT we need to store the actual data in a Datum and be referenced by an NFT. Is it possible to do this in eopsin ? thanks for any pointers.

**Comment by nielstron** [comment, marked as answer]
Not exactly sure what you re trying to achieve with your description (are the two NFTs the same or different ones?)

1) Yes you can mint NTFs with Eopsin (or rather validate the minting). Check out the [universal minting policy example](https://github.com/ImperatorLang/eopsin/blob/universal_minting_policy/examples/smart_contracts/universal_minting_policy.py) and [wrapped token example](https://github.com/ImperatorLang/eopsin/blob/master/examples/smart_contracts/wrapped_token.py)

2) You can also check that some NFT is locked with a specific datum and included in the reference inputs or the normal inputs or outputs. I don't think there are examples yet for checking reference inputs but there are examples for normal inputs/outputs in the [wrapped token example](https://github.com/ImperatorLang/eopsin/blob/master/examples/smart_contracts/wrapped_token.py)

**Reply by reitcircles** [reply]
Our use case is as follows: 
  - we would like to mint an NFT where the metadata should sit at a script address eUtxo (minting script) Datum. And in order to identify which utxo it is, we need to mint an NFT associated with that Utxo. I know this is possible in plutus. Was just exploring how to do this in eopsin.

**Reply by nielstron** [reply]
Unfortunately, based on your description, I neither understand exactly what you are trying to achieve on the high level nor how you want to implement it on the low level.

> I know it can be done in Plutus.

Everything you can do in Plutus should be doable in Eopsin as well (even a bit more). If you want to make it easy to understand what you are trying to do on a low level, feel free to share the corresponding Plutus code.

**Reply by nielstron** [reply]
> we would like to mint an NFT where the metadata should sit at a script address eUtxo (minting script) Datum. And in order to identify which utxo it is, we need to mint an NFT associated with that Utxo. I know this is possible in plutus. Was just exploring how to do this in eopsin.

Here is what I understood so far

1) there is a (script?) address S
2) there is a UTxO U locked at S with associated Datum D (this datum contains metadata of NFT N1?)
3) You want to mint NFT N2 associated (how? is the name the hash of the txid or..?) with U
4) You want to write a minting policy for NFT2 that enforces 3.

There are different ways to go about this, I.e. you can enforce that the minting transaction in 3. also locks the datum D or that it is just a reference input

If I got that right, you would simply write a validator that
1) checks the spending purpose is minting
2) loops through the reference input utxos/ the output utxos to find the utxo with datum D
3) checks that the minted amount of the own policy ID of the nft "associated" with the utxo is 1 and all others 0

**Reply by nielstron** [reply]
Checking names and amounts of minted tokens is demonstrated here: https://github.com/ImperatorLang/eopsin/blob/universal_minting_policy/examples/smart_contracts/universal_minting_policy.py#L34

Looping through inputs/outputs is demonstrated in the wrapped_token sample contract.
https://github.com/ImperatorLang/eopsin/blob/master/examples/smart_contracts/wrapped_token.py#L52
